### ISSUE TITLE
Large multinomial sampling on GPU causes OOM

### ISSUE THREAD

**Issue by mbarnes1** [tensorflow/tensorflow#18058]
Sampling m samples from `tf.multinomial` with support n allocates an unnecessarily large n x m tensor when running on a GPU. For large n and m, this causes out of memory errors. It should be possible to sample from a multinomial distribution without allocating such a large tensor (output tensor is only b x m, where b is the batch size).

### Example and traceback:
```
sample_indices = tf.multinomial(tf.ones([1, 1e6]), 10000, output_dtype=tf.int32)  # sample_indices has shape [1 x 1e4]
with tf.Session() as sess:
    output = sess.run([sample_indices])
```

Traceback:

> ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[1,10000,1000000] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
         [[Node: multinomial/Multinomial = Multinomial[T=DT_FLOAT, output_dtype=DT_INT32, seed=0, seed2=0,
_device="/job:localhost/replica:0/task:0/device:GPU:0"](ones, multinomial/Multinomial/num_samples)]]

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Springdale LInux 7.4 (Redhat)
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: 1.6.0
- **Python version**: 2.7.5
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: CUDA 9.0, cuDNN 7.0.5
- **GPU model and memory**: TITAN X (Pascal) 12GB
- **Exact command to reproduce**: (see above)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cy89** [comment]
@mbarnes1 sorry for the slow reply. Just to check, m and n are 1e6 and 1e4 in your example?

@ebrevdo is this memory behavior intended for tf.multinomial?

**Comment by ebrevdo** [comment]
I believe this is a performance optimization for GPU.  Certain noise is calculated all at once and then combined to generate the multinomial samples.  On the CPU the generator is executed sequentially.  I'm not sure the best way to trade off performance vs. size here.

**Comment by ebrevdo** [comment]
@ekelsen thoughts?  our implementation isn't the most memory efficient.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cy89** [comment]
@zheng-xq can you comment about whether the efficiency of GPU multinomial is on your radar? If not, should we mark this "contributions welcome"?

**Comment by godefv** [comment]
This is also a big issue for me. 

I have: 
batch size=1
number of samples=75520
number of classes=358287

In this context, allocating a tensor of shape (batch size, number of samples, number of classes) is crazy. If there is a trade-off with speed, an alternate function would be welcome, so that the user can choose what he needs.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @mbarnes1 ! 1.x are not supported any more. Can we move this issue to closed status as it is not replicating in [2.8 version](https://github.com/tensorflow/tensorflow/issues/18058)?
